controllers/operatorpolicy_controller.go

```diff
@@ -1426,187 +1470,346 @@ func (r *OperatorPolicyReconciler) handleInstallPlan(
+		if err != nil {
+			return nil, err
+		}
 
-		if sub.Spec.StartingCSV != "" {
-			allowedVersions = append(allowedVersions, policyv1.NonEmptyString(sub.Spec.StartingCSV))
+		if policy == nil {
+			continue
 		}
 
-		for _, acceptableCSV := range allowedVersions {
-			if string(acceptableCSV) == ipCSVs[0] {
-				matchingCSV = true
+		policyTyped := policyv1beta1.OperatorPolicy{}
 
-				break
-			}
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(policy.Object, &policyTyped)
+		if err != nil {
+			continue
 		}
 
-		if matchingCSV {
-			approvedVersion = ipCSVs[0]
+		subTyped := operatorv1alpha1.Subscription{}
+
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(sub.Object, &subTyped)
+		if err != nil {
+			continue
+		}
 
-			approvableInstallPlans = append(approvableInstallPlans, installPlan)
+		approvedCSV := r.getApprovedCSV(&policyTyped, &subTyped, csvNames)
+		if approvedCSV != "" {
+			approvedCSVs.Insert(approvedCSV)
 		}
 	}
 
-	opLog.V(2).Info("Determined approvable InstallPlans", "count", len(approvableInstallPlans))
+	unapprovedCSVs := requiredCSVs.Difference(approvedCSVs).UnsortedList()
 
-	if len(approvableInstallPlans) != 1 {
-		changed := updateStatus(
-			policy,
-			installPlanUpgradeCond(complianceConfig, allUpgradeVersions, approvableInstallPlans),
-			relatedInstallPlans...,
-		)
+	sort.Strings(unapprovedCSVs)
 
-		return changed, nil
+	return unapprovedCSVs, nil
+}
+
+// getApprovedCSV returns the CSV in the passed in csvNames that can be approved. If no approval can take place, an
+// empty string is returned.
+func (r *OperatorPolicyReconciler) getApprovedCSV(
+	policy *policyv1beta1.OperatorPolicy, sub *operatorv1alpha1.Subscription, csvNames []string,
+) string {
+	// Only enforce policies can approve InstallPlans
+	if !policy.Spec.RemediationAction.IsEnforce() {
+		return ""
 	}
 
-	opLog.Info("Approving InstallPlan", "InstallPlanName", approvableInstallPlans[0].GetName(),
-		"InstallPlanNamespace", approvableInstallPlans[0].GetNamespace())
+	initialInstall := sub.Status.InstalledCSV == ""
+	autoUpgrade := policy.Spec.UpgradeApproval == "Automatic"
 
-	if err := unstructured.SetNestedField(approvableInstallPlans[0].Object, true, "spec", "approved"); err != nil {
-		return false, fmt.Errorf("error approving InstallPlan: %w", err)
+	// If the operator is already installed and isn't configured for automatic upgrades, then it can't approve
+	// InstallPlans.
+	if !autoUpgrade && !initialInstall {
+		return ""
 	}
 
-	if err := r.TargetClient.Update(ctx, &approvableInstallPlans[0]); err != nil {
-		return false, fmt.Errorf("error updating approved InstallPlan: %w", err)
+	allowedCSVs := make([]policyv1.NonEmptyString, 0, len(policy.Spec.Versions)+1)
+	allowedCSVs = append(allowedCSVs, policy.Spec.Versions...)
+
+	if sub.Spec.StartingCSV != "" {
+		allowedCSVs = append(allowedCSVs, policyv1.NonEmptyString(sub.Spec.StartingCSV))
+	}
```

Comment:
As a weird note, previously `allowedVersions` was the variable name.

In this review, it is difficult to tell what is new-new code and what is small changes to existing pieces. At least in GH, it's not very good at aligning the old and new... This is obviously not a major issue, but if there were some additional "anchor points" where there weren't changes (these six lines, for example), it might help `git blame` in the future.

Reply:
The side by side diff is easier for this.